### PR TITLE
Add tests for `block_given?` desugaring

### PR DESCRIPTION
**Pull request by amomchilov**
### Motivation

Characterize the behaviour of #9718 before I make some changes.

### Test plan

Just adding new tests specifically to the desugar stage.

### CODE REVIEW

File: test/testdata/desugar/block_given.rb [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+def method_with_anonymous_block_param(&)
+  # Valid cases that we should detect as special calls to `block_given?`
+  block_given?
+  self.block_given?
+  Kernel.block_given?
+  ::Kernel.block_given?
+
+  # Due to the current structure of Desugar.cc, we still detect this as a special `block_given?` call.
+  block_given? { "with a block?!" }
```
Comment (amomchilov) [inline test/testdata/desugar/block_given.rb:30]: Detecting this one very niche edge case would require an extra check for every single block arg, which isn't worth it IMO. It's been fine this long without it.

File: test/testdata/desugar/block_given.rb [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+  #            ^^^^^^^^^^^^^^^^^^^^ error: No body in block
+
+  # Cases we should not detect as special calls to `block_given?`
+  Object.block_given? # Normally not allowed: private method `block_given?' called for Object:Class
+  SomethingElse.block_given?
+  block_given?("with parameter")
+end
+
+def method_with_implicit_block_param()
```
Comment (amomchilov) [inline test/testdata/desugar/block_given.rb:39]: It looks like this `dctx.enclosingBlockParamName.exists()` check was dead code all along:

https://github.com/sorbet/sorbet/blob/0fbadd4c5550e9bc76db624de755c58260d89a8d/ast/desugar/Desugar.cc#L946

By this point in the desugarer, any method that didn't have an explicit block argument, would have had a synthetic `&<blk>` added.

https://github.com/sorbet/sorbet/blob/0fbadd4c5550e9bc76db624de755c58260d89a8d/ast/desugar/Desugar.cc#L375-L382

Comment (amomchilov) [reply test/testdata/desugar/block_given.rb:39]: I'll clean this up in a follow-up

**Comment by amomchilov** [comment]
Mm could you elaborate what you have in mind? @froydnj

**Comment by froydnj** [comment]
> Mm could you elaborate what you have in mind? @froydnj

I was thinking of something like:

```ruby
# typed: true
extend T::Sig

sig {params(blk: T.nilable(T.proc.params(arg0: Integer).void)).void}
def foo(&blk)
  if block_given?
    puts("have a block")
    # I thought this would fail, but Sorbet can see that blk is non-nil here.
    bar(&blk)
  else
    # Sorbet complains, but says blk is nilable here, not nil, which is not great.
    bar(&blk)
    # Ideally this would just fail, because we know blk is nil.
    bar(&T.must(blk))
    puts("no block, sadness")
  end
end

sig {params(blk: T.proc.params(arg0: Integer).void).void}
def bar(&blk)
end
```

Ideally a new desugaring could improve the false case and we would have tests to ensure the true case still worked.

File: test/testdata/desugar/block_given.rb (hunk 2 above)
Comment (amomchilov) [reply test/testdata/desugar/block_given.rb:39]: https://github.com/sorbet/sorbet/pull/9726
